common/web/gesture-recognizer/src/engine/headless/gestureDebugPath.ts

```diff
@@ -0,0 +1,140 @@
+ * - Parameters:
+ *   - `sample: InputSample` - the coordinate & timestamp of the new observation.
+ *
+ * `'complete'`: the touchpoint is no longer active; a touch-end has been observed.
+ *   - Provides no parameters.
+ *   - Will be the last event raised by its instance, after any final 'segmentation'
+ *     events.
+ *   - Still precedes resolution Promise fulfillment on the `Segment` provided by
+ *     the most recently-preceding 'segmentation' event.
+ *     - And possibly recognition Promise fulfillment.
+ *
+ * `'invalidated'`: the touchpoint is no longer active; the path has crossed
+ * gesture-recognition boundaries and is no longer considered valid.
+ *   - Provides no parameters.
+ *   - Will precede the final 'segmentation' event for the 'end' segment
+ *   - Will precede resolution Promise fulfillment on the `Segment` provided by
+ *     the most recently-preceding 'segmentation' event.
+ *     - And possibly recognition Promise fulfillment.
+ */
+export class GestureDebugPath<Type, StateToken = any> extends GesturePath<Type, StateToken> {
+  private samples: InputSample<Type, StateToken>[] = [];
+
+  public clone(): GestureDebugPath<Type, StateToken> {
+    const instance = new GestureDebugPath<Type, StateToken>();
+    instance.samples = [].concat(this.samples);
+
+    instance._isComplete = this._isComplete;
+    instance._wasCancelled = this._wasCancelled;
+    instance._stats = new CumulativePathStats<Type>(this._stats);
+
+    return instance;
+  }
+
+  /**
+   * Deserializes a GesturePath instance from its corresponding JSON.parse() object.
+   * @param jsonObj
+   */
+  static deserialize<Type, StateToken>(jsonObj: SerializedGesturePath<Type, StateToken>): GestureDebugPath<Type, StateToken> {
+    const instance = new GestureDebugPath<Type, StateToken>();
+
+    instance.samples = [].concat(jsonObj.coords.map((obj) => ({...obj} as InputSample<Type, StateToken>)));
+    instance._isComplete = true;
+    instance._wasCancelled = jsonObj.wasCancelled;
+
+    let stats = instance.samples.reduce((stats: CumulativePathStats<Type>, sample) => stats.extend(sample), new CumulativePathStats<Type>());
+    instance._stats = stats;
+
+    return instance;
+  }
+
+  /**
+   * Extends the path with a newly-observed coordinate.
+   * @param sample
+   */
+  extend(sample: InputSample<Type, StateToken>) {
+    /* c8 ignore next 3 */
+    if(this.isComplete) {
+      throw new Error("Invalid state:  this GesturePath has already terminated.");
+    }
+
+    // The tracked path should emit InputSample events before Segment events and
+    // resolution of Segment Promises.
+    this.samples.push(sample);
+    super.extend(sample);
+  }
+
+
+  public translateCoordSystem(functor: (sample: InputSample<Type, StateToken>) => InputSample<Type, StateToken>) {
+    super.translateCoordSystem(functor);
+
+    for(let i=0; i < this.coords.length; i++) {
+      this.samples[i] = functor(this.samples[i]);
+    }
```

Comment:
```suggestion
    for(let i=0; i < this.samples.length; i++) {
      this.samples[i] = functor(this.samples[i]);
    }
```

I know `samples` and `coords` are the same thing but it's still a bit squidgy to have the loop counting off a different variable!